src/expressions.rs

```diff
@@ -30,3 +35,84 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
     })
     .upcast()
 }
+
+pub(crate) fn restriction_expression(chat: &Chat) -> gtk::Expression {
+    let chat_expression = gtk::ConstantExpression::new(chat).upcast();
+    let permissions_expression = chat_expression.chain_property::<Chat>("permissions");
+
+    // chat type is (usually) constant so we don't need to update it
+    match chat.type_() {
+        ChatType::Supergroup(data) if !data.is_channel() => {
+            let member_status_expression = gtk::PropertyExpression::new(
+                Supergroup::static_type(),
+                Some(gtk::ConstantExpression::new(data)),
+                "status",
+            );
+            restriction_label_expression(&member_status_expression, &permissions_expression)
```

Comment:
This better, but can still be improved if you work with type parameters

Here is my suggestion (I also improved the `deep return` spaghetti code)

```rust
pub(crate) fn restriction_expression(chat: &Chat) -> gtk::Expression {
    // chat type is (usually) constant so we don't need to update it
    match chat.type_() {
        ChatType::Supergroup(data) if !data.is_channel() => {
            restriction_label_expression::<Supergroup, _>(data)
        }
        ChatType::BasicGroup(data) => restriction_label_expression::<BasicGroup, _>(data),
        _ => gtk::ConstantExpression::new(&"").upcast(),
    }
}

fn restriction_label_expression<T: glib::StaticType, V: glib::ToValue>(
    value: &V,
) -> gtk::Expression {
    let member_status_expression = gtk::PropertyExpression::new(
        T::static_type(),
        Some(gtk::ConstantExpression::new(value)),
        "status",
    );
    let permissions_expression = Chat::this_expression("permissions");

    gtk::ClosureExpression::new::<String, _, _>(
        &[member_status_expression, permissions_expression],
        closure!(|_: glib::Object, status: BoxedChatMemberStatus, chat_permissions: BoxedChatPermissions| {
            if chat_permissions.0.can_send_messages {
                return String::new()
            }

            match status.0 {
                ChatMemberStatus::Restricted(status) if !status.permissions.can_send_messages => {
                    if status.restricted_until_date == 0 {
                        gettext("The admins of this group have restricted you from writing here")
                    } else {
                        let date =
                            glib::DateTime::from_unix_utc(status.restricted_until_date.into()).unwrap();

                        gettext!(
                            "The admins of this group have restricted you from writing here until {}",
                            date.format(&if glib::DateTime::now_local()
                                .unwrap()
                                .difference(&date)
                                .as_days()
                                == 0
                            {
                                // Translators: This is a hours: minutes time format used in restriction label
                                gettext("%l:%M %p")
                            } else {
                                // Translators: This is a full date format used in restriction label
                                gettext("%B %e, %Y %l:%M %p")
                            })
                            .unwrap()
                            .to_string()
                        )
                    }
                }
                ChatMemberStatus::Creator(_) | ChatMemberStatus::Administrator(_) => {
                    gettext("Writing messages isn't allowed in this group.")
                }
                _ => String::new(),
            }
        }),
    )
    .upcast()
}

```

Reply:
Thanks for your help

Reply:
I believe it should be fine now